sw/host/opentitanlib/src/transport/hyperdebug/i2c.rs

```diff
@@ -0,0 +1,234 @@
+            interface: hyperdebug.i2c_interface,
+            bus_idx: idx,
+            max_read_size: 0x8000 as usize,
+            max_write_size: 0x1000 as usize,
+        })
+    }
+
+    /// Transmit data for a single I2C operation, using one or more USB packets.
+    fn transmit_then_receive(&self, addr: u8, wbuf: &[u8], rbuf: &mut [u8]) -> Result<()> {
+        ensure!(
+            rbuf.len() < self.max_read_size,
+            I2cError::InvalidDataLength(rbuf.len())
+        );
+        ensure!(
+            wbuf.len() < self.max_write_size,
+            I2cError::InvalidDataLength(wbuf.len())
+        );
+        let mut index = if rbuf.len() < 128 {
+            // Short format header
+            let mut req = CmdTransferShort::new();
+            req.port = self.bus_idx | (((wbuf.len() & 0x0F00) >> 4) as u8);
+            req.addr = addr;
+            req.write_count = (wbuf.len() & 0x00FF) as u8;
+            req.read_count = rbuf.len() as u8;
+            let databytes = std::cmp::min(USB_MAX_SIZE - 4, wbuf.len());
+            req.data[0..databytes].clone_from_slice(&wbuf[0..databytes]);
+            self.usb_write_bulk(&req.as_bytes()[0..4 + databytes])?;
+            databytes
+        } else {
+            // Long format header
+            let mut req = CmdTransferLong::new();
+            req.port = self.bus_idx | (((wbuf.len() & 0x0F00) >> 4) as u8);
+            req.addr = addr;
+            req.write_count = (wbuf.len() & 0x00FF) as u8;
+            req.read_count = (rbuf.len() & 0x007F) as u8;
+            req.read_count1 = (rbuf.len() >> 7) as u8;
+            let databytes = std::cmp::min(USB_MAX_SIZE - 6, wbuf.len());
+            req.data[0..databytes].clone_from_slice(&wbuf[0..databytes]);
+            self.usb_write_bulk(&req.as_bytes()[0..6 + databytes])?;
+            databytes
+        };
+
+        // Transmit any more data without further header.
+        while index < wbuf.len() {
+            let databytes = std::cmp::min(USB_MAX_SIZE, wbuf.len() - index);
+            self.usb_write_bulk(&wbuf[index..index + databytes])?;
+            index += databytes;
+        }
+
+        let mut resp = RspTransfer::new();
+        let bytecount = self.usb_read_bulk(&mut resp.as_bytes_mut())?;
+        ensure!(
+            bytecount >= 4,
+            Error::CommunicationError("Unrecognized response to I2C request")
+        );
+        ensure!(
+            resp.status_code == 0,
+            Error::CommunicationError("I2C error")
+        );
+        let databytes = bytecount - 4;
+        rbuf[0..databytes].clone_from_slice(&resp.data[0..databytes]);
+        let mut index = databytes;
+        while index < rbuf.len() {
+            let databytes = self.usb_read_bulk(&mut resp.data[index..])?;
+            ensure!(
+                databytes > 0,
+                Error::CommunicationError("Unrecognized reponse to I2C request")
+            );
+            index += databytes;
```

Comment:
A somewhat more idiomatic way to write this might be..
```rust
let mut buf = &mut resp.data[..];
while !buf.is_empty() {
  let len = self.usb_read_bulk(buf);
  buf = &mut buf[len..];
}
```

Reply:
In the similar USB protocol for SPI, the current byte index was needed for headers on each USB packet.  It is true that as that is not the case for this I2C protocol, I could rewrite as you suggest, but on the other hand I would like to keep the two implementation similar if possible.